Hamoye Internship Stage A data analysis

In [ ]:
#import necessary libraries
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display # display

#data table formatting
from google.colab import data_table
#data_table.enable_dataframe_formatter()

#to disable it 
data_table.disable_dataframe_formatter()

%matplotlib inline

In [ ]:
#Setting display
pd.set_option('display.max_columns', 50) # Setting pandas to display a N number of columns
pd.set_option('display.max_rows', 30) # Setting pandas to display a N number rows
pd.set_option('display.width', 500) # Setting pandas dataframe display width to N

In [ ]:
#For automated visualization

#uncomment the code to install the libraries
#!pip install --upgrade autoviz pandas-profiling

#import the libraries
import pandas_profiling as pp # library for automatic EDA
from autoviz.AutoViz_Class import AutoViz_Class

In [ ]:
#First step is to load the dataset 
data_url = "https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv"

#load csv file into data variable
data = pd.read_csv(data_url)

#using pd.Dataframe.head() to 
#preview the first 5 rows of the 
#dataset
data.head()

record_id  utility_id_ferc1  report_year      plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu
0   f1_fuel_1994_12_1_0_7                 1         1994              rockport                coal       ton        5377489.0               16.590                      18.59                         18.53                1.121
1  f1_fuel_1994_12_1_0_10                 1         1994  rockport total plant                coal       ton       10486945.0               16.592                      18.58                         18.53                1.120
2   f1_fuel_1994_12_2_0_1                 2         1994                gorgas                coal       ton        2978683.0               24.130                      39.72                         38.12                1.650
3   f1_fuel_1994_12_2_0_7                 2         1994                 barry                coal       ton        3739484.0               23.950                      47.21                         45.99                1.970
4  f1_fuel_1994_12_2_0_10                 2         1994             chickasaw                 gas       mcf          40533.0                1.000                       2.77                          2.77                2.570

In [ ]:
#using pandas describe function to
#get a summary of the dataset
data.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu
count      29523.000000  29523.000000     2.952300e+04         29523.000000               29523.000000                  2.952300e+04         29523.000000
mean         118.601836   2005.806050     2.622119e+06             8.492111                 208.649031                  9.175704e+02            19.304354
std           74.178353      7.025483     9.118004e+06            10.600220                2854.490090                  6.877593e+04          2091.540939
min            1.000000   1994.000000     1.000000e+00             0.000001                -276.080000                 -8.749370e+02           -41.501000
25%           55.000000   2000.000000     1.381700e+04             1.024000                   5.207000                  3.778500e+00             1.940000
50%          122.000000   2006.000000     2.533220e+05             5.762694                  26.000000                  1.737100e+01             4.127000
75%          176.000000   2012.000000     1.424034e+06            17.006000                  47.113000                  4.213700e+01             7.745000
max          514.000000   2018.000000     5.558942e+08           341.260000              139358.000000                  7.964521e+06        359278.000000

It was observed that the dataset contains 29523 rows of data on fuel consumption.

In [ ]:
#using the info function to further 
#gain understanding of the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29523 entries, 0 to 29522
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   record_id                     29523 non-null  object 
 1   utility_id_ferc1              29523 non-null  int64  
 2   report_year                   29523 non-null  int64  
 3   plant_name_ferc1              29523 non-null  object 
 4   fuel_type_code_pudl           29523 non-null  object 
 5   fuel_unit                     29343 non-null  object 
 6   fuel_qty_burned               29523 non-null  float64
 7   fuel_mmbtu_per_unit           29523 non-null  float64
 8   fuel_cost_per_unit_burned     29523 non-null  float64
 9   fuel_cost_per_unit_delivered  29523 non-null  float64
 10  fuel_cost_per_mmbtu           29523 non-null  float64
dtypes: float64(5), int64(2), object(4)
memory usage: 2.5+ MB


In [ ]:
#checking for missing number
data.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

Sometimes, the missing number which is equivalent to np.nan (Nan is not a number) could fail to capture data like '?' etc. So we need to validate that the missing numbers are actually 180 rows for fuel unit.

In [ ]:
missing_count = 0
for values in data.values.tolist():
    for value in values:
        if value == '?':
            missing_count += 1
        else:
            continue
#Print the result
print('We have {} missing values not captured by isna/isnull command'.format(missing_count))

We have 0 missing values not captured by isna/isnull command


So the missing values were all captured by np.nan which was integrated into pandas.isna() call

In [ ]:
#fuel unit has few missing numbers
#we can check the fraction of the missing number to the total data
print(data.fuel_unit.isna().sum()/len(data.fuel_unit) *100)

0.609694136774718


From the result, it was noted that the amount of missing rows is so low and removing the affected rows won't affect our analysis.

Secondly, replacing the values would either influence our findings positively or negatively.

So let's drop them for now into a new variable called data_df

In [ ]:
data_df = data.dropna()
data_df

record_id  utility_id_ferc1  report_year      plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu
0        f1_fuel_1994_12_1_0_7                 1         1994              rockport                coal       ton        5377489.0               16.590                      18.59                         18.53                1.121
1       f1_fuel_1994_12_1_0_10                 1         1994  rockport total plant                coal       ton       10486945.0               16.592                      18.58                         18.53                1.120
2        f1_fuel_1994_12_2_0_1                 2         1994                gorgas                coal       ton        2978683.0               24.130                      39.72                         38.12                1.650
3        f1_fuel_1994_12_2_0_7                 2         1994                 barry                coal       ton        3739484.0               23.950                      47.21                         45.99                1.970
4       f1_fuel_1994_12_2_0_10                 2         1994             chickasaw                 gas       mcf          40533.0                1.000                       2.77                          2.77                2.570
...                        ...               ...          ...                   ...                 ...       ...              ...                  ...                        ...                           ...                  ...
29518  f1_fuel_2018_12_12_0_13                12         2018    neil simpson ct #1                 gas       mcf          18799.0                1.059                       4.78                          4.78                9.030
29519   f1_fuel_2018_12_12_1_1                12         2018  cheyenne prairie 58%                 gas       mcf         806730.0                1.050                       3.65                          3.65                6.950
29520  f1_fuel_2018_12_12_1_10                12         2018     lange ct facility                 gas       mcf         104554.0                1.060                       4.77                          4.77                8.990
29521  f1_fuel_2018_12_12_1_13                12         2018       wygen 3 bhp 52%                coal       ton         315945.0               16.108                       3.06                         14.76                1.110
29522  f1_fuel_2018_12_12_1_14                12         2018       wygen 3 bhp 52%                 gas       mcf          17853.0                1.059                       0.00                          0.00               11.680

[29343 rows x 11 columns]

In [ ]:
#now to verify the drop
display(data_df.isna().sum())

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

In [ ]:
#lets check the columns of the dataset
for column in data_df.columns:
    display()
    print("Column name:", column)
    print("Datatype", data_df[column].dtypes)
    print()

Column name: record_id
Datatype object

Column name: utility_id_ferc1
Datatype int64

Column name: report_year
Datatype int64

Column name: plant_name_ferc1
Datatype object

Column name: fuel_type_code_pudl
Datatype object

Column name: fuel_unit
Datatype object

Column name: fuel_qty_burned
Datatype float64

Column name: fuel_mmbtu_per_unit
Datatype float64

Column name: fuel_cost_per_unit_burned
Datatype float64

Column name: fuel_cost_per_unit_delivered
Datatype float64

Column name: fuel_cost_per_mmbtu
Datatype float64



Focusing on the columns that holds information that we can work with.
Let carryout out some visualization of the data

There are two ways to do this.
1. Automated visualization using either pandas profiling or autoviz 

2. Using Matplotlib and seaborn 

So let's start with number 2 first

In [ ]:
#Since we have imported Matplotlib and seaborn
#let's plot each column against each other.
#aside record id and utility id 

#storing all the features in the dataset
features = list(data_df.columns[2:])

In [ ]:
#preview the features we need
for column in features:
    display()
    print(column)

report_year
plant_name_ferc1
fuel_type_code_pudl
fuel_unit
fuel_qty_burned
fuel_mmbtu_per_unit
fuel_cost_per_unit_burned
fuel_cost_per_unit_delivered
fuel_cost_per_mmbtu


In [ ]:
#let check the datatypes again
for column in features:
    print(column)
    display(data_df[column].dtypes)
    print()

report_year


dtype('int64')


plant_name_ferc1


dtype('O')


fuel_type_code_pudl


dtype('O')


fuel_unit


dtype('O')


fuel_qty_burned


dtype('float64')


fuel_mmbtu_per_unit


dtype('float64')


fuel_cost_per_unit_burned


dtype('float64')


fuel_cost_per_unit_delivered


dtype('float64')


fuel_cost_per_mmbtu


dtype('float64')

Where 'O' is Object

In [ ]:
#before plotting, let's check the 
#number of unique values on each
#column

for col in features:
    print(col)
    print(data_df[col].nunique())
    print()

report_year
25

plant_name_ferc1
2307

fuel_type_code_pudl
6

fuel_unit
9

fuel_qty_burned
26270

fuel_mmbtu_per_unit
11178

fuel_cost_per_unit_burned
19335

fuel_cost_per_unit_delivered
16612

fuel_cost_per_mmbtu
12565



So on the x axis, we can only use fuel unit and fuel type code and peradventure manage to use record year.
Because the unique values are low, unlike others with high unique values, we will fit them into y axis when plotting them against x axis

In [ ]:
#Lets take the x axis columns.
#First let's get their index position
#in the feature lists
for x in features:
    print(x)
    print(features.index(x))
    print()

report_year
0

plant_name_ferc1
1

fuel_type_code_pudl
2

fuel_unit
3

fuel_qty_burned
4

fuel_mmbtu_per_unit
5

fuel_cost_per_unit_burned
6

fuel_cost_per_unit_delivered
7

fuel_cost_per_mmbtu
8



In [33]:
#using function to select the column
x, y = [], []
for col in features:
    if data_df[col].nunique() <= 30:
        x.append(col)
    else:
        y.append(col)

So we to take index 0, 2, and 3

In [ ]:
#i would be using a function to visualize the plots
x2 = features[2:4] #report year
x2.append(features[0])
y2 = features [4:]
y2.append(features[1])

In [34]:
print(x)
print()
print(x2)

['report_year', 'fuel_type_code_pudl', 'fuel_unit']

['fuel_type_code_pudl', 'fuel_unit', 'report_year']


In [ ]:
y

['plant_name_ferc1',
 'fuel_qty_burned',
 'fuel_mmbtu_per_unit',
 'fuel_cost_per_unit_burned',
 'fuel_cost_per_unit_delivered',
 'fuel_cost_per_mmbtu']

In [ ]:
#using pairplot
sns.set()
for col in x:
    sns.pairplot(data_df[:1000], hue=col, size=2.5);
    plt.title(col.capitalize())

In [ ]:
# plotting with factorplot
with sns.axes_style(style='ticks'):
    for col_x in x:
        plt.title(col.capitalize())
        for col_y in y:
            if data_df[col_y].dtypes != 'object':
                g = sns.factorplot(col_x, col_y, data=data_df[:1000], size=5);
            g.set_axis_labels(col_x, col_y);
            print("\n"*5) #to give space between plots
        print()

In [ ]:
#using the automated visualization

pp_report = pp.ProfileReport(data_df)
pp_report

In [35]:
#using Autoviz

import nltk
nltk.download('punkt')
nltk.download('wordnet')

#initiate the class
Autoviz = AutoViz_Class()

#visualize the dataset by passing
#in the data url or csv file
Autoviz.AutoViz(data_url)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
Shape of your Data Set loaded: (29523, 11)
############## C L A S S I F Y I N G  V A R I A B L E S  ####################
Classifying variables in data set...
    11 Predictors classified...
        1 variables removed since they were ID or low-information variables
Number of All Scatter Plots = 15
Time to run AutoViz = 29 seconds 

 ###################### AUTO VISUALIZATION Completed ########################


record_id  utility_id_ferc1  report_year      plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu
0        f1_fuel_1994_12_1_0_7                 1         1994              rockport                coal       ton        5377489.0               16.590                      18.59                         18.53                1.121
1       f1_fuel_1994_12_1_0_10                 1         1994  rockport total plant                coal       ton       10486945.0               16.592                      18.58                         18.53                1.120
2        f1_fuel_1994_12_2_0_1                 2         1994                gorgas                coal       ton        2978683.0               24.130                      39.72                         38.12                1.650
3        f1_fuel_1994_12_2_0_7                 2         1994                 barry                coal       ton        3739484.0               23.950                      47.21                         45.99                1.970
4       f1_fuel_1994_12_2_0_10                 2         1994             chickasaw                 gas       mcf          40533.0                1.000                       2.77                          2.77                2.570
...                        ...               ...          ...                   ...                 ...       ...              ...                  ...                        ...                           ...                  ...
29518  f1_fuel_2018_12_12_0_13                12         2018    neil simpson ct #1                 gas       mcf          18799.0                1.059                       4.78                          4.78                9.030
29519   f1_fuel_2018_12_12_1_1                12         2018  cheyenne prairie 58%                 gas       mcf         806730.0                1.050                       3.65                          3.65                6.950
29520  f1_fuel_2018_12_12_1_10                12         2018     lange ct facility                 gas       mcf         104554.0                1.060                       4.77                          4.77                8.990
29521  f1_fuel_2018_12_12_1_13                12         2018       wygen 3 bhp 52%                coal       ton         315945.0               16.108                       3.06                         14.76                1.110
29522  f1_fuel_2018_12_12_1_14                12         2018       wygen 3 bhp 52%                 gas       mcf          17853.0                1.059                       0.00                          0.00               11.680

[29523 rows x 11 columns]

For any visualization that didn't work, try running the codes again.

Thanks for reading

Cheers

Miracle Osigwe
ID: 147de030a4c1f000